[TableGen] HasOneUse builtin predicate on PatFrags

jofrn opened the pull request:
This predicate tells GlobalISelEmitter and DAGISelEmitter to check that the instruction to emit has only one use of its result. This can be used on a PatFrag instead of defining custom predicates for both emitters per record that requires it.

arsenm commented on llvm/include/llvm/CodeGen/GlobalISel/GIMatchTableExecutorImpl.h:
```diff
@@ -468,7 +468,23 @@ bool GIMatchTableExecutor::executeMatchTable(
+      const MachineInstr *MI = State.MIs[InsnID];
+      assert(MI && "Used insn before defined");
+      assert(MI->getNumDefs() > 0 && "No defs");
+      const Register Res = MI->getOperand(0).getReg();
```
We might want an extended variant that can handle multiple defs at some point

Pierre-vh replied:
Yes, it depends on how many of these are emitted at once for a given pattern, we need to see more use first
Though this is just 2 bytes on average (1 byte opcode + 1 byte id) so it's not a big deal unless this is emitted in bulk in a ton of rules

arsenm commented on llvm/test/TableGen/predicate-patfags.td:
```diff
@@ -32,10 +34,15 @@ def : GINodeEquiv<G_TGT_MUL24, TGTmul24_impl>;
+#ifndef HASONEUSE
   [{ return N->hasOneUse(); }]> {
   let GISelPredicateCode = [{
     return MRI->hasOneNonDBGUse(MI.getOperand(0).getReg());
```
Maybe should change the sample predicate code

arsenm reviewed (commented):
Should also switch the AMDGPU HasOneUseUnaryOp/HasOneUseBinOp/HasOneUseTernaryOp to use this

arsenm commented:
> > Should also switch the AMDGPU HasOneUseUnaryOp/HasOneUseBinOp/HasOneUseTernaryOp to use this
> 
> Right, thanks. I was thinking that these changes can go in a follow-up commit. The current change is a general one for [TableGen], but the next one can be [AMDGPU] with the actual usages of the `HasOneUse` predicate on these `HasOneUse{Unary,Bin,Ternary}Op`s.

Can do it either way. At least it's more clear it works with the real uses

Pierre-vh commented on llvm/include/llvm/CodeGen/GlobalISel/GIMatchTableExecutorImpl.h:
```diff
@@ -468,7 +468,23 @@ bool GIMatchTableExecutor::executeMatchTable(
+      assert(MI->getNumDefs() > 0 && "No defs");
+      const Register Res = MI->getOperand(0).getReg();
+
+      if (!MRI.hasOneNonDBGUse(Res))
+        if (handleReject() == RejectAndGiveUp)
+          return false;
```
```suggestion
      if (!MRI.hasOneNonDBGUse(Res)) {
        if (handleReject() == RejectAndGiveUp)
          return false;
      }
```

Pierre-vh commented on llvm/utils/TableGen/GlobalISelEmitter.cpp:
```diff
@@ -210,6 +210,9 @@ static Error isTrivialOperatorNode(const TreePatternNode &N) {
     if (Predicate.hasNoUse())
       continue;
 
+    if (Predicate.hasOneUse())
```
Merge those with `||` ?

jofrn committed:
[TableGen] HasOneUse builtin predicate on PatFrags

This predicate tells GlobalISelEmitter and DAGISelEmitter to check that the instruction to emit has only one use of its result. This can be used on a PatFrag instead of defining custom predicates for both emitters per record that requires it.
